src/server/input/input_event_transformer.cpp

```diff
@@ -116,28 +116,34 @@ mir::input::InputEventTransformer::~InputEventTransformer() = default;
 
 bool mi::InputEventTransformer::transform(MirEvent const& event)
 {
-    std::lock_guard lock{mutex};
-
-    for (auto it = input_transformers.begin(); it != input_transformers.end();)
+    // This cache needs to be introduced because a transformer may be removed in response
+    // to one of its own dispatched events (e.g. a transformer may dispatch some sort of
+    // click that removes the transformer itself). As a result, we cannot be holding the
+    // lock to the transformer while it is trying to remove itself.
+    std::vector<std::shared_ptr<Transformer>> transformers_cache;
+    transformers_cache.reserve(input_transformers.size());
     {
-        if (it->expired())
-        {
-            it = input_transformers.erase(it);
-        }
-        else
+        std::lock_guard lock{mutex};
```

Comment:
Looks like we're accessing `input_transformers.size()` and then guarding access to `input_transformers`